apps/api/v2/src/modules/webhooks/guards/is-oauth-client-webhook-guard.ts

```diff
@@ -0,0 +1,49 @@
+import { GetUserReturnType } from "@/modules/auth/decorators/get-user/get-user.decorator";
+import { OAuthClientRepository } from "@/modules/oauth-clients/oauth-client.repository";
+import { WebhooksService } from "@/modules/webhooks/services/webhooks.service";
+import { CanActivate, ExecutionContext, Injectable, NotFoundException } from "@nestjs/common";
+import { Request } from "express";
+
+import { PlatformOAuthClient, Webhook } from "@calcom/prisma/client";
+
+@Injectable()
+export class IsOAuthClientWebhookGuard implements CanActivate {
+  constructor(
+    private readonly webhooksService: WebhooksService,
+    private readonly oAuthClientRepository: OAuthClientRepository
+  ) {}
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const request = context
+      .switchToHttp()
+      .getRequest<Request & { webhook: Webhook; oAuthClient: PlatformOAuthClient }>();
+    const user = request.user as GetUserReturnType;
+    const webhookId = request.params.webhookId;
+    const oAuthClientId = request.params.clientId;
+
+    if (!user || !webhookId || !oAuthClientId) {
+      return false;
+    }
+
+    const oAuthClient = await this.oAuthClientRepository.getOAuthClient(oAuthClientId);
+
+    if (!oAuthClient) {
+      throw new NotFoundException(`OAuthClient (${oAuthClientId}) not found`);
+    }
+
+    const webhook = await this.webhooksService.getWebhookById(webhookId);
+
+    if (oAuthClient?.organizationId !== user.movedToProfile?.organizationId) {
+      return user.isSystemAdmin;
+    }
+
+    if (webhook.platformOAuthClientId !== oAuthClientId) {
+      return false;
```

Comment:
Just like on line 25 and 37 we return false - I am thinking if its best to throw unauthenticated 400 manually with a unique error message, because as we have more and more guards if on line 25 we return false its harder to track what and why is not passing.

Also, I would break down on line 24 for each failed state aka !user and !webhookId separate error messages.